File: BinaryTreeRightSideView/binary_tree_right_side_view.py
#!/usr/bin/env python3

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        from collections import deque

        q = deque()
        if root: q.append(root)
        
        res = []
        while q:
            rightmost = None

            for _ in range(len(q)):
                node = q.popleft()
                rightmost = node

                if node.left: q.append(node.left)
                if node.right: q.append(node.right)

            res.append(rightmost.val)
        return res

File: BinaryTreeRightSideView/binary_tree_right_side_view.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> res {};

        deque<TreeNode*> q {};
        if ( root ) q.push_back( root );

        while ( !q.empty() ){
            TreeNode* rightmost;
            size_t len {q.size()};

            for ( size_t i {0}; i != len; i++ ){
                TreeNode* node {q.front()};
                q.pop_front();
                rightmost = node;
                if ( node->left ) q.push_back(node->left);
                if ( node->right ) q.push_back(node->right);
            }
            res.push_back( rightmost->val );
        }
        return res;
    }
};

